Initial implementation of streaming metric

Dref360 opened the pull request:
Summary:

In #655, we propose streaming metrics so that we can use Fairlearn metrics with batches as it is common in Deep learning. The naive implementation is to simply accumulate the values before computing the metric when asked.

**NOTE** This is WIP, open to suggestions and I wanted some guidance to be sure I didn't miss something obvious. I will add unit tests before making it ready for review.

Related issues
* Related to #655 (does not fully fix it).

Quick test:

```python
import numpy as np
import sklearn.metrics as skm
from fairlearn.metrics import MetricFrame
classes = [0,1]
sensitive = ['a', 'b']
y_true = np.random.choice(classes, [100], replace=True)
y_pred = np.random.choice(classes, [100], replace=True)
sensitive_attrs = np.random.choice(sensitive, [100], replace=True)
met = MetricFrame(skm.recall_score, y_true=y_true, y_pred=y_pred, sensitive_features=sensitive_attrs)
met.overall


# Batched, we simulate 2 batches.
yt1,yt2 = y_true[:40], y_true[40:]
yp1,yp2 = y_pred[:40], y_pred[40:]
sa1,sa2 = sensitive_attrs[:40], sensitive_attrs[40:]


met_strt = MetricFrame(skm.recall_score, y_true=[], y_pred=[], sensitive_features=[], streaming=True)
met_strt.add_data(yt1, yp1, sensitive_features=sa1)
met_strt.add_data(yt2, yp2, sensitive_features=sa2)

# Check that it's equal.
assert met_strt.overall == met.overall
```

Signed-off-by: Dref360 <frederic.branchaud-charron@usherbrooke.ca>

MiroDudik commented on fairlearn/metrics/_metric_frame.py:
```diff
@@ -117,24 +126,127 @@ def __init__(self,
+                 y_pred, *,
+                 sensitive_features,
+                 control_features: Optional = None,
+                 ):
```
Do we want to add `sample_params` so as to pretty much follow the same format as constructor minus `metric`?

Dref360 replied:
I think this would lead to errors.
Example:
```
met = MetricFrame(skm.recall_score, y_true=[], y_pred=[], sensitive_features=[], streaming=True, sampling_params={'average':'micro'})

met_strt.add_data(..., sample_params={'average':'macro'})
....
met_strt.add_data(..., sample_params={'average':'micro'})
```

Then we compute the value differently for different batches.

riedgar-ms commented on fairlearn/metrics/_metric_frame.py:
```diff
@@ -226,6 +338,9 @@ def overall(self) -> Union[Any, pd.Series, pd.DataFrame]:
             interface when calling programatically, while also reducing
             typing for those using Fairlearn interactively.
         """
+        if self._overall is None:
```
Does this mean that once a user has invoked one of these summary functions, they can no longer keep streaming data into the object?

Dref360 replied:
We can still keep it, but we will wipe `_overall` and `by_group` so the next call will recompute it.

riedgar-ms replied:
Ahh yes, I see above.

riedgar-ms reviewed (commented):
Could you start adding some tests, please? The sample code in the PR description is helpful, but it's much better to have something which is running in the automation.

Dref360 commented on fairlearn/metrics/_metric_frame.py:
```diff
@@ -117,24 +126,127 @@ def __init__(self,
-        func_dict = self._process_functions(metric, sample_params)
+        self._y_t.append(y_t)
+        self._y_p.append(y_p)
+        self._s_f.append(sensitive_features)
```
I will add support for more formats, but these accumulators accumulate batches and we concatenate them when needed.

Dref360 commented:
I will add tests later this week.

MiroDudik commented on fairlearn/metrics/_metric_frame.py:
```diff
@@ -117,24 +126,127 @@ def __init__(self,
         """Read a placeholder comment."""
+        if len(y_true) == 0 and not streaming:
+            raise ValueError(_BAD_INIT_ERROR)
+
```
Folks, I'd like to hit a pause and open a quick discussion about what this should look like internally. My suggestion is that we should create an intermediate `DataFrame` representation. Specifically:

**1. We create an internal function of the following form:**

```python
def _pandify(y_true, y_pred, sensitive_features, control_features, sample_params) -> pd.DataFrame:
    
    # My sample code below assumes that y_pred and y_true are 1D array-like,
    # sensitive_features and control_features are dicts of 1D array-like,
    # sample_params are dicts of dicts of 1D array-like.
    # 
    # The actual code would first need to do validation and conversion
    # including the generation of names of control features and sensitive features.
    
    shared_inputs_df = pd.concat(
        {'dummy': pd.DataFrame({'y_true': y_true, 'y_pred': y_pred})}, axis=1)
    sensitive_features_df = pd.concat(
        {'dummy': pd.DataFrame(sensitive_features)}, axis=1)     
    control_features_df = pd.concat(
        {'dummy': pd.DataFrame(control_features)}, axis=1)           
    sample_params_df = pd.concat(
        {key: pd.DataFrame(sample_params[key]) for key in sample_params.keys()}, axis=1)

    combined_df = pd.concat({'shared_inputs': shared_inputs_df,
        'sensitive_features': sensitive_features_df,
        'control_features': control_features_df, 
        'sample_params': sample_params_df}, axis=1)
    return combined_df
```

**2. For non-streaming implementation:**
We call `_pandify` in `__init__` and then calculate `self._overall` and `self._by_group`.

**3. For streaming implementation:**
We call `_pandify` in `add_data`. If this is not the first call, we check that the columns of the new data frame agree with the columns of the old data frame and if they do, we call `pd.concat` on them.

What do you think? (The reason for my micromanaging this is that there are a few extensions of `MetricFrame` in the pipeline and I think they'll be much smoother if we have the intermediate `DataFrame` representation.)

MiroDudik replied:
There would be some updating in other internal processing functions--we would no longer be passing multiple arguments, e.g., we would have
```python
def _compute_overall(self, func_dict, combined_df):
```

Dref360 replied:
That is a great idea! It would simplify a lot of things :) I'll work on it this week.

MiroDudik replied:
I can see that the internal representation becomes more obscure, but I don't understand the column consistency issue. If we have the current data in `df` and new data in `df_new` we can just check consistency by calling `df.columns.symmetric_difference(df_new.columns).empty` and we can just update with `df = df.append(df_new)`.

adrinjalali replied:
These operations you're mentioning are quite expensive. And in terms of API design, it's much easier to understand what the input to a method is, if it gets things it needs in separate arguments compared to accepting a single dataframe which the user of the method would need to know how to prepare.

riedgar-ms replied:
Is this issue still outstanding? It does seem quite major based on @adrinjalali 's comments.

adrinjalali replied:
If I'm not mistaken, we are NOT going down the pandas way, so I'd say we're good?

adrinjalali commented on fairlearn/metrics/_metric_frame.py:
```diff
@@ -117,24 +126,127 @@ def __init__(self,
+        if not self._streaming:
+            raise Exception("This MetricFrame does not support adding data.")
         check_consistent_length(y_true, y_pred)
+        check_consistent_length(y_true, sensitive_features)
```
I _think_ you can pass multiple arrays to the same function call here

adrinjalali reviewed (commented):
Here's a few notes :) Thanks @Dref360

Dref360 commented:
I added some tests and implemented most of the comments. Thank you everyone!

As for _pandify, @MiroDudik I'm not sure how to handle "metric-wide" sample_params like `average` in `skm.recall_score` using this method. I would need some guidance.

riedgar-ms commented on fairlearn/metrics/_metric_frame.py:
```diff
@@ -147,8 +288,9 @@ def __init__(self,
 
-        self._overall = self._compute_overall(func_dict, y_t, y_p, cf_list)
-        self._by_group = self._compute_by_group(func_dict, y_t, y_p, sf_list, cf_list)
+        self.func_dict = self._process_functions(self.metric, s_p)
```
Why has the processing of the functions been moved out of the constructor?

Dref360 replied:
Because the process_functions is dependent on the sample_params.

riedgar-ms commented on test/unit/metrics/test_metricframe_streaming.py:
```diff
@@ -0,0 +1,131 @@
+    target = metrics.MetricFrame(skm.recall_score, [], [], sensitive_features=[],
+                                 sample_params=None, streaming=True)
+
+    with pytest.raises(ValueError) as excinfo:
```
I think this should be a separate test.

riedgar-ms replied:
And on a related point, what happens if you call `.overall` before calling `add_data()` when streaming?

Dref360 replied:
It will raise `ValueError('No data to process.')` Exception message to be reworded.

riedgar-ms commented on test/unit/metrics/test_metricframe_streaming.py:
```diff
@@ -0,0 +1,131 @@
+@pytest.mark.parametrize("transform_y_p", conversions_for_1d)
+@pytest.mark.parametrize("transform_y_t", conversions_for_1d)
+@pytest.mark.parametrize("num_batches", [1, 3, 5])
+def test_basic_streaming(transform_y_t, transform_y_p, num_batches):
```
Thanks for adding these. Could you also have a test where the batches change type each time? e.g. the first `add_data()` call uses a list, the next a Series, the next an ndarray etc.?

Dref360 replied:
Hmm my current implementation does not support it. I'll think on how to support it.

riedgar-ms replied:
I think we should

riedgar-ms replied:
Have you added these tests? I seem to be overlooking them.

riedgar-ms commented on test/unit/metrics/test_metricframe_streaming.py:
```diff
@@ -0,0 +1,131 @@
+    assert isinstance(target.by_group, pd.Series)
+
+    # Check we have expected number of elements
+    assert len(target.by_group) == 2
```
Given that the non-streaming tests already do this sort of checking, would it be easier to just put the data into a non-streaming `MetricFrame` and then check for equality?

Dref360 replied:
Right!

riedgar-ms commented on test/unit/metrics/test_metricframe_streaming.py:
```diff
@@ -0,0 +1,131 @@
+@pytest.mark.parametrize("num_batches", [1, 3, 5])
+def test_basic_streaming(transform_y_t, transform_y_p, num_batches):
+    batch_generator = batchify(num_batches, y_t, y_p, g_4)
+    target = metrics.MetricFrame(skm.recall_score, [], [], sensitive_features=[],
```
Could we also have a case where some subgroups of the sensitive feature(s) are not present in the initial batch?

riedgar-ms replied:
And another case where there are intermixed calls to `add_data()` and `.overall` and `.by_group` etc.?

Dref360 commented:
Hello!
Sorry for the delay.
I think I answered most questions and fixed most issues. So I would welcome a new round of review.

Of course, if we want to change the design I am open to suggestions!

riedgar-ms commented on fairlearn/metrics/_metric_frame.py:
```diff
@@ -117,24 +129,156 @@ def __init__(self,
+            self._compute_metric(y_t, y_p, sensitive_features, control_features, sample_params)
+        else:
+            # If we are streaming, we wait, but initialize accumulators.
+            assert y_true == [] and y_pred == [] and sensitive_features == [], _NON_EMPTY_INIT_ERR
```
This will eventually need to get changed to a ValueError, but that can wait until we're on to the polishing stage.

Dref360 replied:
done

riedgar-ms commented on fairlearn/metrics/_metric_frame.py:
```diff
@@ -117,24 +129,156 @@ def __init__(self,
+            metric function name, with the values being the string-to-array-like dictionaries.
+        """
+        if not self._streaming:
+            raise Exception("This MetricFrame does not support adding data.")
```
Again, not vital quite yet, but could we raise a more specific exception? Although I'm not sure that `ValueError` is appropriate here.

Dref360 replied:
done

riedgar-ms replied:
@adrinjalali  would have better thoughts than I do on that.

adrinjalali replied:
Isn't this a `NotImplementedError`?

riedgar-ms commented on fairlearn/metrics/_metric_frame.py:
```diff
@@ -117,24 +129,156 @@ def __init__(self,
+        """
+        if not self._streaming:
+            raise Exception("This MetricFrame does not support adding data.")
+        check_consistent_length(y_true, y_pred, sensitive_features)
```
Shouldn't there be a check on `sample_params` for consistent lengths as well?

riedgar-ms replied:
Not to mention `control_features`

riedgar-ms replied:
Checked below, I see.

riedgar-ms commented on fairlearn/metrics/_metric_frame.py:
```diff
@@ -117,24 +129,156 @@ def __init__(self,
+            elif isinstance(self._s_p, list):
+                self._s_p.append(sample_params)
+            else:
+                raise ValueError("MetricFrame was initialized with `sample_params` already set.")
```
I don't understand this error. Aren't there two cases which should be being caught:
- User provides `sample_params` on the first call to this method, but then forgets on a subsequent call
- User _doesn't_ provide `sample_params` on the first call, but tries to on some later call

Dref360 replied:
There is also the case where the user supplies `sample_params` at initialization. For example, to provide `average='binary'` to `skm.f1_score`. Then, the user is not allowed to supply `sample_params` after.

riedgar-ms replied:
Surely `average='binary'` isn't a _sample_ parameter? By definition, sample parameters have a unique value for each sample - things like sample weights.

riedgar-ms reviewed (commented):
Thanks for keeping at this!

riedgar-ms commented on test/unit/metrics/test_metricframe_streaming.py:
```diff
@@ -148,3 +154,72 @@ def test_exception_on_no_data():
+
+    batch_generator = batchify(num_batches, y_t, y_p, g_4)
+    # We initialize with CF
+    target = metrics.MetricFrame(skm.recall_score, [], [], sensitive_features=[],
```
Can you make this 'success' case a separate test?

riedgar-ms commented on test/unit/metrics/test_metricframe_streaming.py:
```diff
@@ -148,3 +154,72 @@ def test_exception_on_no_data():
+
+    # Version with sample_weights supplied at the start.
+    target = metrics.MetricFrame(skm.recall_score, [], [], sensitive_features=[],
+                                 sample_params={'sample_weight': sample_weights},
```
Shouldn't this be `sample_params={'sample_weight': [] }` here - the sample weights would need to be streamed as well?

riedgar-ms reviewed (commented):
Sorry for the delay in reviewing.

riedgar-ms commented:
@Dref360 did you have any questions about my review comments? Have I missed something else you've changed?

Dref360 commented:
Sorry about that, I'll work on the comments this week.

riedgar-ms commented:
> Sorry about that, I'll work on the comments this week.

No problem - we know this is volunteer work!

riedgar-ms commented on fairlearn/metrics/_metric_frame.py:
```diff
@@ -119,24 +133,162 @@ def __init__(self,
                  y_pred, *,
```
As a warning #822 is changing this signature slightly. There may be merge conflicts.

riedgar-ms commented on fairlearn/metrics/_metric_frame.py:
```diff
@@ -119,24 +133,162 @@ def __init__(self,
+            result = sum(batches, [])
+        elif batch_type is pd.DataFrame:
+            col_nums = batches[0].columns
+            assert all([col_nums == df.columns for df in batches]), 'Need same columns'
```
Can this raise a `ValueError` too?

riedgar-ms commented on test/unit/metrics/test_metricframe_streaming.py:
```diff
@@ -0,0 +1,215 @@
+    target = metrics.MetricFrame(skm.recall_score, [], [], sensitive_features=[],
+                                 streaming=True)
+    # We supply CF in the call, it should fail.
+    with pytest.raises(ValueError):
```
Should there be a check on the exception message, so you can be sure that it raised on the right statement?

riedgar-ms commented on fairlearn/metrics/_metric_frame.py:
```diff
@@ -119,24 +133,162 @@ def __init__(self,
+        self._overall = None
+        self._by_group = None
+
+    def _concat_batches(self, batches):
```
I know that it's technically private, but it might be worth writing tests of this specific method. It may be easier to make sure that each branch is checked that way.

adrinjalali replied:
we really need to have that "patch" check @riedgar-ms , it'd be easier to review here then.

riedgar-ms replied:
Do you mean the code coverage check?

adrinjalali replied:
yes

riedgar-ms reviewed (commented):
Thank you for keeping at this!

riedgar-ms reviewed (commented):
Tagging @adrinjalali  and @MiroDudik  ... @Dref360 deserves a break from my reviews :-)

adrinjalali commented on fairlearn/metrics/_metric_frame.py:
```diff
@@ -195,24 +209,166 @@ def __init__(self,
+        self._by_group = None
+
+    def _concat_batches(self, batches):
+        """Concatenate a list of items together."""
```
please explain in the docstring what this function's supposed to do exactly, the code seems a bit not so nice here.

adrinjalali commented on fairlearn/metrics/_metric_frame.py:
```diff
@@ -195,24 +209,166 @@ def __init__(self,
+    def _concat_batches(self, batches):
+        """Concatenate a list of items together."""
+        if len(batches) == 0:
+            raise ValueError(_EMPTY_BATCHES_ERR)
```
instead of raise, I'd just return None.

Dref360 replied:
I'm afraid that this could lead to issues in the future.
Other libraries also raise:

Numpy: `ValueError: need at least one array to concatenate`
Pandas: `ValueError: No objects to concatenate`

riedgar-ms replied:
I'm happy with raising here.

adrinjalali commented on fairlearn/metrics/_metric_frame.py:
```diff
@@ -195,24 +209,166 @@ def __init__(self,
+    def _compute_metric(self, y_t, y_p, s_f, c_f, s_p):
         # Now, prepare the sensitive features
-        sf_list = self._process_features("sensitive_feature_", sensitive_features, y_t)
+        sf_list = self._process_features("sensitive_feature_", s_f, y_t)
```
could we please not have the very shortened version of the variable names?

adrinjalali reviewed (commented):
Thanks for all the work so far @Dref360 , it's certainly getting there.

riedgar-ms commented on fairlearn/metrics/_metric_frame.py:
```diff
@@ -171,6 +179,13 @@ class MetricFrame:
+        the score. One can append data with ``add_data``. The user can't supply data in the
+        constructor if ``streaming=True``.
+
+        **Note** currently, it only stores the values before computing the metric.
```
I'm not quite sure that this note is supposed to mean.

Dref360 replied:
In #655, we mentioned that in the future we would like to support object that have an "update function. Like [Keras Stateful Metrics](https://keras.io/api/metrics/) or [BaaL Metrics](https://github.com/ElementAI/baal/blob/master/src/baal/utils/metrics.py#L22).

In this PR, we simply accumulate the predictions and we run the metrics on all values.

riedgar-ms replied:
Isn't that always going to be required?

Dref360 replied:
Yeah I guess, unless we have multiple metrics object at the same time for each combination of sensitive attribute.

I'll remove for now and we can decide what to do later.

riedgar-ms commented on fairlearn/metrics/_metric_frame.py:
```diff
@@ -195,24 +210,176 @@ def __init__(self,
+            accumulators = [y_true, y_pred, sensitive_features]
+            # CF must be None or an empty list.
+            cf_validated = control_features is None or len(control_features) == 0
+            # Same for SP
```
Nit: can we write out 'SP' and 'CF' in the comment at least.

riedgar-ms commented on test/unit/metrics/test_metricframe_streaming.py:
```diff
@@ -0,0 +1,255 @@
+    assert target._overall is None  # has been reset
+    assert target._by_group is None  # has been reset
+
+    # We can recompute it with the new batches.
```
You mean that invoking `by_group` implicitly recomputes?

riedgar-ms commented on test/unit/metrics/test_metricframe_streaming.py:
```diff
@@ -0,0 +1,255 @@
+    assert target._by_group is None
+
+    # We can't initialize streaming metrics with data
+    with pytest.raises(ValueError) as excinfo:
```
This looks like it should be a separate test.

riedgar-ms reviewed (commented):
Thank you for the update!

adrinjalali commented on fairlearn/metrics/_metric_frame.py:
```diff
@@ -3,32 +3,40 @@
-    "Keys in 'sample_params' do not match those in 'metric'"
+_SAMPLE_PARAM_KEYS_NOT_IN_FUNC_DICT = ("Keys in 'sample_params'"
+                                       " do not match those in 'metric'")
+_BAD_INIT_ERROR = """Non streaming metrics require data to process.
+ Set streaming=True for streaming metrics"""
```
we should be consistent, either have newlines in all messages or none, I prefer none.

adrinjalali commented on fairlearn/metrics/_metric_frame.py:
```diff
@@ -194,24 +207,181 @@ def __init__(self,
+            raise NotImplementedError(_NOT_STREAMING_ERR)
+        check_consistent_length(y_true, y_pred, sensitive_features)
+
         y_t = _convert_to_ndarray_and_squeeze(y_true)
         y_p = _convert_to_ndarray_and_squeeze(y_pred)
```
@riedgar-ms how does this work with the arbitrary y_true, y_pred shapes as you have it in the example we just merged?

riedgar-ms replied:
Aren't these lines unchanged? Now, it's true that doing the squeeze might have unpleasant effects if the user supplied metric functions were expecting the extra dimensions. I'm not sure that's an issue for this PR, though.

adrinjalali commented on fairlearn/metrics/_metric_frame.py:
```diff
@@ -194,24 +207,181 @@ def __init__(self,
+            cf_validated = control_features is None or len(control_features) == 0
+            # Same for sample_params
+            sp_validator = sample_params in (None, {})
+            if any(acc != [] for acc in accumulators) or not cf_validated or not sp_validator:
```
```suggestion
            if any(acc for acc in accumulators) or not cf_validated or not sp_validator:
```

adrinjalali commented on fairlearn/metrics/_metric_frame.py:
```diff
@@ -194,24 +207,181 @@ def __init__(self,
+
+        Parameters
+        ----------
+        y_true : List, pandas.Series, numpy.ndarray, pandas.DataFrame
```
```suggestion
        y_true : array-like
```

?

romanlutz replied:
Doesn't sphinx complain about this since it's not technically a "type"?

adrinjalali replied:
I don't think so: https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.FastICA.html#sklearn.decomposition.FastICA.fit_transform

adrinjalali replied:
This is the guidelines sklearn has: https://scikit-learn.org/stable/developers/contributing.html?highlight=contribute#guidelines-for-writing-documentation

adrinjalali commented on fairlearn/metrics/_metric_frame.py:
```diff
@@ -194,24 +207,181 @@ def __init__(self,
+        y_pred : List, pandas.Series, numpy.ndarray, pandas.DataFrame
+                The predictions.
+
+        sensitive_features : List, pandas.Series, dict of 1d arrays, numpy.ndarray,
+         pandas.DataFrame
```
```suggestion
        sensitive_features : array-like, or dict of array-like
```

?

adrinjalali commented on fairlearn/metrics/_metric_frame.py:
```diff
@@ -194,24 +207,181 @@ def __init__(self,
+        if not all([type(arr) is batch_type for arr in batches]):
+            raise ValueError("Can't concatenate arrays of different types.")
+        # Simple cases, array, Series and list can be concat easily.
+        if batch_type is np.ndarray:
+            result = np.concatenate(batches)
+        elif batch_type is list:
+            result = sum(batches, [])
+        elif batch_type is pd.Series:
+            result = pd.concat(batches)
+        elif batch_type is pd.DataFrame:
+            # To concat dataframe, we check if the columns match.
+            col_nums = batches[0].columns
+            if any([col_nums != df.columns for df in batches]):
+                raise ValueError(f"Column mismatch expected {col_nums},"
+                                 f" got {[df.columns for df in batches]}")
+            result = pd.concat(batches)
+        elif batch_type is dict:
+            # Create a new dict that merges values of all keys.
+            all_keys = set(sum((list(b.keys()) for b in batches), []))
+            return {k: self._concat_batches([b[k] for b in batches if k in b])
+                    for k in all_keys}
+        else:
+            raise ValueError(f"Can't concatenate {batches}")
```
I think I'd be happier if we just convert everything to numpy arrays and concatenate them instead of all these different cases. Do we need them to stay the same type?

adrinjalali reviewed (commented):
Thanks for all the work so far @Dref360 

One general concern I have is that this is touching a lot of the MetricFrame. The question is, do we want it to be the same class, or rather have a `StreamingMetricFrame` where everything is just assumed to be streamed? We can share the common code in a common parent class.

We should also add this to our user guides.

adrinjalali commented:
Maybe with a bit of refactoring I'd be happier with having only one class, but right now reviewing the PR gives me the impression that it's not very maintainable.

Dref360 commented:
I agree that simply making a wrapper class around MetricFrame would be easier, so I'll close this PR and open another later with a new implementation.

Thank you for your reviews!
